core/metrics-core-impl/src/main/java/org/hawkular/metrics/core/impl/MetricsServiceImpl.java

```diff
@@ -455,6 +457,94 @@ public void setTaskScheduler(TaskScheduler taskScheduler) {
+        // currently write a separate batch statement for each metric.
+        //
+        // TODO Is there additional overhead of using batch statement when there is only a single insert?
+        //      If there is overhead, then we should avoid using batch statements when the metric has only a single
+        //      data point which could be quite often.
+        //
+        // The metrics_idx table stores the metric id along with any tags, and data retention. The update we perform
+        // here though only inserts the metric id (i.e., name and interval). We need to revisit this logic. The original
+        // intent for updating metrics_idx here is that even if the client does not explicitly create the metric, we
+        // still have it in metrics_idx. In reality, I think clients will be explicitly creating metrics. This will
+        // certainly be the case with the full, integrated hawkular server.
+        //
+        // TODO Determine how much overhead is caused by updating metrics_idx on every write
+        //      If there much overhead, then we might not want to update the index every time we insert data. Maybe
+        //      we periodically update it in the background, so we will still be aware of metrics that have not been
+        //      explicitly created, just not necessarily right away.
+
+        PublishSubject<Void> results = PublishSubject.create();
+        Observable<Integer> updates = metrics
+                .filter(metric -> !metric.getDataPoints().isEmpty())
+                .flatMap(metric -> {
+                    switch (metric.getId().getType()) {
+                        case GAUGE:
+                        case COUNTER_RATE:
+                            @SuppressWarnings("unchecked")
+                            Metric<Double> gauge = (Metric<Double>) metric;
+                            return dataAccess.insertGaugeData(gauge, getTTL(gauge.getId()))
+                                    .doOnNext(gaugeInserts::mark);
+                        case AVAILABILITY:
+                            @SuppressWarnings("unchecked")
+                            Metric<AvailabilityType> avail = (Metric<AvailabilityType>) metric;
+                            return dataAccess.insertAvailabilityData(avail, getTTL(avail.getId()))
+                                    .doOnNext(availabilityInserts::mark);
+                        case COUNTER:
+                            @SuppressWarnings("unchecked")
+                            Metric<Long> counter = (Metric<Long>) metric;
+                            return dataAccess.insertCounterData(counter, getTTL(counter.getId()))
+                                    .doOnNext(counterInserts::mark);
+                        default:
+                            throw new IllegalArgumentException(metric.getId().getType().getText());
+                    }
+                });
+        Observable<Integer> indexUpdates = dataAccess.updateMetricsIndex(metrics);
```

Comment:
Le 19/08/2015 23:26, Michael Burman a écrit :
> Behavior is not identical to previous version here. On purpose? Not to

I've compiled the three methods into one (only #addAvailData filtered on 
empty data points).

And I had to move updates of the Meter instances along with the calls to 
DataAccess#insertXXXData

While it does not behave exactly the same internally, the result is 
comparable, isn't it ?


> mention like I discussed with @jsanda <https://github.com/jsanda> we
> should get rid of the PublishSubject in this one (and the extra
> subscribe) as it resubscribes to the source observable.

I wasn't aware of this discussion. There's no mention of it in JIRA or 
the devel list. Can you update the ticket?